tools/bitfield_gen.md

```diff
@@ -0,0 +1,653 @@
+<!--
+  Copyright 2023 Proofcraft Pty Ltd
+  SPDX-License-Identifier: CC-BY-SA-4.0
+-->
+
+# Bitfield Generator Manual
+
+## Introduction
+
+The tool `bitfield_gen.py` in the `seL4/tools/` directory generates bitfield
+accessor code in C with formal specifications and proofs in Isabelle/HOL.
+
+It takes specifications of bitfield layouts in memory such as the following
+
+```bf_gen
+block VMFault {
+    field     address           32
+
+    field     FSR               5
+    padding                     7
+    field     instructionFault  1
+    padding                     15
+    field     seL4_FaultType    4
+}
+```
+
+and generates a C struct with constructor and getter/setter functions with
+call-by-value and call-by-pointer semantics. It also, separately, generates a
+corresponding data type definition in Isabelle HOL with corresponding functions
+in HOL for that data type, as well as proofs that the generated C code correctly
+implements these functions. This includes absence of undefined behaviour in C.
```

Comment:
It might be useful to explain why this tool was made. My understanding is that it's because the C standard doesn't completely specify how bit-fields are laid out, and originally the compiler implementations tended to be inconsistent or have bugs?

Reply:
One other thought is that by knowing and controlling the exact bitfield allocations, the fastpath implementations are allowed to rely on layout optimizations.

Reply:
Good point. I'll leave out the compiler bugs from back when on gcc for Arm, because I suspect they have been fixed by now and that is mostly of historical interest, but the precise layout semantics are still needed over standard C bitfields.